// @mrleebo/prisma-ast/dist/prisma-ast.esm.js
import { createToken, Lexer, CstParser } from 'chevrotain';
import { lilconfigSync } from 'lilconfig';
import { EOL } from 'os';

const Identifier = /*#__PURE__*/createToken({
  name: 'Identifier',
  pattern: /[a-zA-Z][\w-]*/
});
const True = /*#__PURE__*/createToken({
  name: 'True',
  pattern: /true/,
  longer_alt: Identifier
});
const False = /*#__PURE__*/createToken({
  name: 'False',
  pattern: /false/,
  longer_alt: Identifier
});
const Null = /*#__PURE__*/createToken({
  name: 'Null',
  pattern: /null/,
  longer_alt: Identifier
});
const Comment = /*#__PURE__*/createToken({
  name: 'Comment',
  pattern: Lexer.NA
});
const DocComment = /*#__PURE__*/createToken({
  name: 'DocComment',
  pattern: /\/\/\/[ \t]*(.*)/,
  categories: [Comment]
});
const LineComment = /*#__PURE__*/createToken({
  name: 'LineComment',
  pattern: /\/\/[ \t]*(.*)/,
  categories: [Comment]
});
const Attribute = /*#__PURE__*/createToken({
  name: 'Attribute',
  pattern: Lexer.NA
});
const BlockAttribute = /*#__PURE__*/createToken({
  name: 'BlockAttribute',
  pattern: /@@/,
  label: "'@@'",
  categories: [Attribute]
});
const FieldAttribute = /*#__PURE__*/createToken({
  name: 'FieldAttribute',
  pattern: /@/,
  label: "'@'",
  categories: [Attribute]
});
const Dot = /*#__PURE__*/createToken({
  name: 'Dot',
  pattern: /\./,
  label: "'.'"
});
const QuestionMark = /*#__PURE__*/createToken({
  name: 'QuestionMark',
  pattern: /\?/,
  label: "'?'"
});
const LCurly = /*#__PURE__*/createToken({
  name: 'LCurly',
  pattern: /{/,
  label: "'{'"
});
const RCurly = /*#__PURE__*/createToken({
  name: 'RCurly',
  pattern: /}/,
  label: "'}'"
});
const LRound = /*#__PURE__*/createToken({
  name: 'LRound',
  pattern: /\(/,
  label: "'('"
});
const RRound = /*#__PURE__*/createToken({
  name: 'RRound',
  pattern: /\)/,
  label: "')'"
});
const LSquare = /*#__PURE__*/createToken({
  name: 'LSquare',
  pattern: /\[/,
  label: "'['"
});
const RSquare = /*#__PURE__*/createToken({
  name: 'RSquare',
  pattern: /\]/,
  label: "']'"
});
const Comma = /*#__PURE__*/createToken({
  name: 'Comma',
  pattern: /,/,
  label: "','"
});
const Colon = /*#__PURE__*/createToken({
  name: 'Colon',
  pattern: /:/,
  label: "':'"
});
const Equals = /*#__PURE__*/createToken({
  name: 'Equals',
  pattern: /=/,
  label: "'='"
});
const StringLiteral = /*#__PURE__*/createToken({
  name: 'StringLiteral',
  pattern: /"(:?[^\\"\n\r]|\\(:?[bfnrtv"\\/]|u[0-9a-fA-F]{4}))*"/
});
const NumberLiteral = /*#__PURE__*/createToken({
  name: 'NumberLiteral',
  pattern: /-?(0|[1-9]\d*)(\.\d+)?([eE][+-]?\d+)?/
});
const WhiteSpace = /*#__PURE__*/createToken({
  name: 'WhiteSpace',
  pattern: /\s+/,
  group: Lexer.SKIPPED
});
const LineBreak = /*#__PURE__*/createToken({
  name: 'LineBreak',
  pattern: /\n|\r\n/,
  line_breaks: true,
  label: 'LineBreak'
});
const tokens = [Comment, DocComment, LineComment, LineBreak, WhiteSpace, Attribute, BlockAttribute, FieldAttribute, Dot, QuestionMark, LCurly, RCurly, LSquare, RSquare, LRound, RRound, Comma, Colon, Equals, True, False, Null, StringLiteral, NumberLiteral, Identifier];
const PrismaLexer = /*#__PURE__*/new Lexer(tokens);

const schemaObjects = ['model', 'view', 'type'];
function isOneOfSchemaObjects(obj, schemas) {
  return obj != null && 'type' in obj && schemas.includes(obj.type);
}
function isSchemaObject(obj) {
  return isOneOfSchemaObjects(obj, schemaObjects);
}
const fieldObjects = ['field', 'enumerator'];
function isSchemaField(field) {
  return field != null && 'type' in field && fieldObjects.includes(field.type);
}
function isToken(node) {
  return 'image' in node[0];
}
function appendLocationData(data, ...tokens) {
  const location = tokens.reduce((memo, token) => {
    if (!token) return memo;
    const {
      endColumn = -Infinity,
      endLine = -Infinity,
      endOffset = -Infinity,
      startColumn = Infinity,
      startLine = Infinity,
      startOffset = Infinity
    } = memo;
    if (token.startLine != null && token.startLine < startLine) memo.startLine = token.startLine;
    if (token.startColumn != null && token.startColumn < startColumn) memo.startColumn = token.startColumn;
    if (token.startOffset != null && token.startOffset < startOffset) memo.startOffset = token.startOffset;
    if (token.endLine != null && token.endLine > endLine) memo.endLine = token.endLine;
    if (token.endColumn != null && token.endColumn > endColumn) memo.endColumn = token.endColumn;
    if (token.endOffset != null && token.endOffset > endOffset) memo.endOffset = token.endOffset;
    return memo;
  }, {});
  return Object.assign(data, {
    location
  });
}

const defaultConfig = {
  parser: {
    nodeLocationTracking: 'none'
  }
};
let config;
function getConfig() {
  if (config != null) return config;
  const result = lilconfigSync('prisma-ast').search();
  return config = Object.assign(defaultConfig, result?.config);
}

class PrismaParser extends CstParser {
  constructor(config) {
    super(tokens, config);
    this.config = void 0;
    this.break = this.RULE('break', () => {
      this.CONSUME1(LineBreak);
      this.CONSUME2(LineBreak);
    });
    this.keyedArg = this.RULE('keyedArg', () => {
      this.CONSUME(Identifier, {
        LABEL: 'keyName'
      });
      this.CONSUME(Colon);
      this.SUBRULE(this.value);
    });
    this.array = this.RULE('array', () => {
      this.CONSUME(LSquare);
      this.MANY_SEP({
        SEP: Comma,
        DEF: () => {
          this.SUBRULE(this.value);
        }
      });
      this.CONSUME(RSquare);
    });
    this.object = this.RULE('object', () => {
      this.CONSUME(LCurly);
      this.MANY_SEP({
        SEP: Comma,
        DEF: () => {
          this.SUBRULE(this.keyedArg);
        }
      });
      this.CONSUME(RCurly);
    });
    this.func = this.RULE('func', () => {
      this.CONSUME(Identifier, {
        LABEL: 'funcName'
      });
      this.CONSUME(LRound);
      this.MANY_SEP({
        SEP: Comma,
        DEF: () => {
          this.OR([{
            ALT: () => this.SUBRULE(this.keyedArg)
          }, {
            ALT: () => this.SUBRULE(this.value)
          }]);
        }
      });
      this.CONSUME(RRound);
    });
    this.value = this.RULE('value', () => {
      this.OR([{
        ALT: () => this.CONSUME(StringLiteral, {
          LABEL: 'value'
        })
      }, {
        ALT: () => this.CONSUME(NumberLiteral, {
          LABEL: 'value'
        })
      }, {
        ALT: () => this.SUBRULE(this.array, {
          LABEL: 'value'
        })
      }, {
        ALT: () => this.SUBRULE(this.object, {
          LABEL: 'value'
        })
      }, {
        ALT: () => this.SUBRULE(this.func, {
          LABEL: 'value'
        })
      }, {
        ALT: () => this.CONSUME(True, {
          LABEL: 'value'
        })
      }, {
        ALT: () => this.CONSUME(False, {
          LABEL: 'value'
        })
      }, {
        ALT: () => this.CONSUME(Null, {
          LABEL: 'value'
        })
      }, {
        ALT: () => this.CONSUME(Identifier, {
          LABEL: 'value'
        })
      }]);
    });
    this.property = this.RULE('property', () => {
      this.CONSUME(Identifier, {
        LABEL: 'propertyName'
      });
      this.CONSUME(Equals);
      this.SUBRULE(this.value, {
        LABEL: 'propertyValue'
      });
    });
    this.assignment = this.RULE('assignment', () => {
      this.CONSUME(Identifier, {
        LABEL: 'assignmentName'
      });
      this.CONSUME(Equals);
      this.SUBRULE(this.value, {
        LABEL: 'assignmentValue'
      });
    });
    this.field = this.RULE('field', () => {
      this.CONSUME(Identifier, {
        LABEL: 'fieldName'
      });
      this.SUBRULE(this.value, {
        LABEL: 'fieldType'
      });
      this.OPTION1(() => {
        this.OR([{
          ALT: () => {
            this.CONSUME(LSquare, {
              LABEL: 'array'
            });
            this.CONSUME(RSquare, {
              LABEL: 'array'
            });
          }
        }, {
          ALT: () => this.CONSUME(QuestionMark, {
            LABEL: 'optional'
          })
        }]);
      });
      this.MANY(() => {
        this.SUBRULE(this.fieldAttribute, {
          LABEL: 'attributeList'
        });
      });
      this.OPTION2(() => {
        this.CONSUME(Comment, {
          LABEL: 'comment'
        });
      });
    });
    this.block = this.RULE('block', (options = {}) => {
      const {
        componentType
      } = options;
      const isEnum = componentType === 'enum';
      const isObject = componentType === 'model' || componentType === 'view' || componentType === 'type';
      this.CONSUME(LCurly);
      this.CONSUME1(LineBreak);
      this.MANY(() => {
        this.OR([{
          ALT: () => this.SUBRULE(this.comment, {
            LABEL: 'list'
          })
        }, {
          GATE: () => isObject,
          ALT: () => this.SUBRULE(this.property, {
            LABEL: 'list'
          })
        }, {
          ALT: () => this.SUBRULE(this.blockAttribute, {
            LABEL: 'list'
          })
        }, {
          GATE: () => isObject,
          ALT: () => this.SUBRULE(this.field, {
            LABEL: 'list'
          })
        }, {
          GATE: () => isEnum,
          ALT: () => this.SUBRULE(this.enum, {
            LABEL: 'list'
          })
        }, {
          GATE: () => !isObject,
          ALT: () => this.SUBRULE(this.assignment, {
            LABEL: 'list'
          })
        }, {
          ALT: () => this.SUBRULE(this.break, {
            LABEL: 'list'
          })
        }, {
          ALT: () => this.CONSUME2(LineBreak)
        }]);
      });
      this.CONSUME(RCurly);
    });
    this.enum = this.RULE('enum', () => {
      this.CONSUME(Identifier, {
        LABEL: 'enumName'
      });
      this.MANY(() => {
        this.SUBRULE(this.fieldAttribute, {
          LABEL: 'attributeList'
        });
      });
      this.OPTION(() => {
        this.CONSUME(Comment, {
          LABEL: 'comment'
        });
      });
    });
    this.fieldAttribute = this.RULE('fieldAttribute', () => {
      this.CONSUME(FieldAttribute, {
        LABEL: 'fieldAttribute'
      });
      this.OR([{
        ALT: () => {
          this.CONSUME1(Identifier, {
            LABEL: 'groupName'
          });
          this.CONSUME(Dot);
          this.CONSUME2(Identifier, {
            LABEL: 'attributeName'
          });
        }
      }, {
        ALT: () => this.CONSUME(Identifier, {
          LABEL: 'attributeName'
        })
      }]);
      this.OPTION(() => {
        this.CONSUME(LRound);
        this.MANY_SEP({
          SEP: Comma,
          DEF: () => {
            this.SUBRULE(this.attributeArg);
          }
        });
        this.CONSUME(RRound);
      });
    });
    this.blockAttribute = this.RULE('blockAttribute', () => {
      this.CONSUME(BlockAttribute, {
        LABEL: 'blockAttribute'
      }), this.OR([{
        ALT: () => {
          this.CONSUME1(Identifier, {
            LABEL: 'groupName'
          });
          this.CONSUME(Dot);
          this.CONSUME2(Identifier, {
            LABEL: 'attributeName'
          });
        }
      }, {
        ALT: () => this.CONSUME(Identifier, {
          LABEL: 'attributeName'
        })
      }]);
      this.OPTION(() => {
        this.CONSUME(LRound);
        this.MANY_SEP({
          SEP: Comma,
          DEF: () => {
            this.SUBRULE(this.attributeArg);
          }
        });
        this.CONSUME(RRound);
      });
    });
    this.attributeArg = this.RULE('attributeArg', () => {
      this.OR([{
        ALT: () => this.SUBRULE(this.keyedArg, {
          LABEL: 'value'
        })
      }, {
        ALT: () => this.SUBRULE(this.value, {
          LABEL: 'value'
        })
      }]);
    });
    this.component = this.RULE('component', () => {
      const type = this.CONSUME1(Identifier, {
        LABEL: 'type'
      });
      this.OR2([{
        ALT: () => {
          this.CONSUME2(Identifier, {
            LABEL: 'groupName'
          });
          this.CONSUME(Dot);
          this.CONSUME3(Identifier, {
            LABEL: 'componentName'
          });
        }
      }, {
        ALT: () => this.CONSUME4(Identifier, {
          LABEL: 'componentName'
        })
      }]);
      this.SUBRULE(this.block, {
        ARGS: [{
          componentType: type.image
        }]
      });
    });
    this.comment = this.RULE('comment', () => {
      this.CONSUME(Comment, {
        LABEL: 'text'
      });
    });
    this.schema = this.RULE('schema', () => {
      this.MANY(() => {
        this.OR([{
          ALT: () => this.SUBRULE(this.comment, {
            LABEL: 'list'
          })
        }, {
          ALT: () => this.SUBRULE(this.component, {
            LABEL: 'list'
          })
        }, {
          ALT: () => this.SUBRULE(this.break, {
            LABEL: 'list'
          })
        }, {
          ALT: () => this.CONSUME(LineBreak)
        }]);
      });
    });
    this.performSelfAnalysis();
    this.config = config;
  }
}
const defaultParser = /*#__PURE__*/new PrismaParser(getConfig().parser);

const VisitorClassFactory = parser => {
  const BasePrismaVisitor = parser.getBaseCstVisitorConstructorWithDefaults();
  return class PrismaVisitor extends BasePrismaVisitor {
    constructor() {
      super();
      this.validateVisitor();
    }
    schema(ctx) {
      const list = ctx.list?.map(item => this.visit([item])) || [];
      return {
        type: 'schema',
        list
      };
    }
    component(ctx) {
      const [type] = ctx.type;
      const [name] = ctx.componentName;
      const list = this.visit(ctx.block);
      const data = (() => {
        switch (type.image) {
          case 'datasource':
            return {
              type: 'datasource',
              name: name.image,
              assignments: list
            };
          case 'generator':
            return {
              type: 'generator',
              name: name.image,
              assignments: list
            };
          case 'model':
            return {
              type: 'model',
              name: name.image,
              properties: list
            };
          case 'view':
            return {
              type: 'view',
              name: name.image,
              properties: list
            };
          case 'enum':
            return {
              type: 'enum',
              name: name.image,
              enumerators: list
            };
          case 'type':
            return {
              type: 'type',
              name: name.image,
              properties: list
            };
          default:
            throw new Error(`Unexpected block type: ${type}`);
        }
      })();
      return this.maybeAppendLocationData(data, type, name);
    }
    break() {
      return {
        type: 'break'
      };
    }
    comment(ctx) {
      const [comment] = ctx.text;
      const data = {
        type: 'comment',
        text: comment.image
      };
      return this.maybeAppendLocationData(data, comment);
    }
    block(ctx) {
      return ctx.list?.map(item => this.visit([item]));
    }
    assignment(ctx) {
      const value = this.visit(ctx.assignmentValue);
      const [key] = ctx.assignmentName;
      const data = {
        type: 'assignment',
        key: key.image,
        value
      };
      return this.maybeAppendLocationData(data, key);
    }
    field(ctx) {
      const fieldType = this.visit(ctx.fieldType);
      const [name] = ctx.fieldName;
      const attributes = ctx.attributeList?.map(item => this.visit([item]));
      const comment = ctx.comment?.[0]?.image;
      const data = {
        type: 'field',
        name: name.image,
        fieldType,
        array: ctx.array != null,
        optional: ctx.optional != null,
        attributes,
        comment
      };
      return this.maybeAppendLocationData(data, name, ctx.optional?.[0], ctx.array?.[0]);
    }
    fieldAttribute(ctx) {
      const [name] = ctx.attributeName;
      const [group] = ctx.groupName || [{}];
      const args = ctx.attributeArg?.map(attr => this.visit(attr));
      const data = {
        type: 'attribute',
        name: name.image,
        kind: 'field',
        group: group.image,
        args
      };
      return this.maybeAppendLocationData(data, name, ...ctx.fieldAttribute, group);
    }
    blockAttribute(ctx) {
      const [name] = ctx.attributeName;
      const [group] = ctx.groupName || [{}];
      const args = ctx.attributeArg?.map(attr => this.visit(attr));
      const data = {
        type: 'attribute',
        name: name.image,
        kind: 'object',
        group: group.image,
        args
      };
      return this.maybeAppendLocationData(data, name, ...ctx.blockAttribute, group);
    }
    attributeArg(ctx) {
      const value = this.visit(ctx.value);
      return {
        type: 'attributeArgument',
        value
      };
    }
    func(ctx) {
      const [name] = ctx.funcName;
      const params = ctx.value?.map(item => this.visit([item]));
      const keyedParams = ctx.keyedArg?.map(item => this.visit([item]));
      const pars = (params || keyedParams) && [...(params ?? []), ...(keyedParams ?? [])];
      const data = {
        type: 'function',
        name: name.image,
        params: pars
      };
      return this.maybeAppendLocationData(data, name);
    }
    array(ctx) {
      const args = ctx.value?.map(item => this.visit([item]));
      return {
        type: 'array',
        args
      };
    }
    object(ctx) {
      const properties = ctx.keyedArg?.map(item => this.visit([item])) ?? [];
      return {
        type: 'object',
        properties
      };
    }
    keyedArg(ctx) {
      const [key] = ctx.keyName;
      const value = this.visit(ctx.value);
      const data = {
        type: 'keyValue',
        key: key.image,
        value
      };
      return this.maybeAppendLocationData(data, key);
    }
    value(ctx) {
      if (isToken(ctx.value)) {
        const [{
          image
        }] = ctx.value;
        return image;
      }
      return this.visit(ctx.value);
    }
    enum(ctx) {
      const [name] = ctx.enumName;
      const attributes = ctx.attributeList?.map(item => this.visit([item]));
      const comment = ctx.comment?.[0]?.image;
      const data = {
        type: 'enumerator',
        name: name.image,
        attributes,
        comment
      };
      return this.maybeAppendLocationData(data, name);
    }
    maybeAppendLocationData(data, ...tokens) {
      if (parser.config.nodeLocationTracking === 'none') return data;
      return appendLocationData(data, ...tokens);
    }
  };
};
const DefaultVisitorClass = /*#__PURE__*/VisitorClassFactory(defaultParser);
const defaultVisitor = /*#__PURE__*/new DefaultVisitorClass();

function getSchema(source, options) {
  const lexingResult = PrismaLexer.tokenize(source);
  const parser = options?.parser ?? defaultParser;
  parser.input = lexingResult.tokens;
  const cstNode = parser.schema();
  if (parser.errors.length > 0) throw parser.errors[0];
  const visitor = options?.visitor ?? defaultVisitor;
  return visitor.visit(cstNode);
}

const unsorted = ['break', 'comment'];
const defaultSortOrder = ['generator', 'datasource', 'model', 'view', 'enum', 'break', 'comment'];
const schemaSorter = (schema, locales, sortOrder = defaultSortOrder) => (a, b) => {
  const aUnsorted = unsorted.indexOf(a.type) !== -1;
  const bUnsorted = unsorted.indexOf(b.type) !== -1;
  if (aUnsorted !== bUnsorted) {
    return schema.list.indexOf(a) - schema.list.indexOf(b);
  }
  if (sortOrder !== defaultSortOrder) sortOrder = sortOrder.concat(defaultSortOrder);
  const typeIndex = sortOrder.indexOf(a.type) - sortOrder.indexOf(b.type);
  if (typeIndex !== 0) return typeIndex;
  if ('name' in a && 'name' in b) return a.name.localeCompare(b.name, locales);
  return 0;
};

function printSchema(schema, options = {}) {
  const {
    sort = false,
    locales = undefined,
    sortOrder = undefined
  } = options;
  let blocks = schema.list;
  if (sort) {
    blocks = schema.list = blocks.filter(block => block.type !== 'break');
    const sorter = schemaSorter(schema, locales, sortOrder);
    blocks.sort(sorter);
  }
  return blocks.map(printBlock).filter(Boolean).join(EOL).replace(/(\r?\n\s*){3,}/g, EOL + EOL) + EOL;
}
function printBlock(block) {
  switch (block.type) {
    case 'comment':
      return printComment(block);
    case 'datasource':
      return printDatasource(block);
    case 'enum':
      return printEnum(block);
    case 'generator':
      return printGenerator(block);
    case 'model':
    case 'view':
    case 'type':
      return printObject(block);
    case 'break':
      return printBreak();
    default:
      throw new Error(`Unrecognized block type`);
  }
}
function printComment(comment) {
  return comment.text;
}
function printBreak() {
  return EOL;
}
function printDatasource(db) {
  const children = computeAssignmentFormatting(db.assignments);
  return `
datasource ${db.name} {
  ${children}
}`;
}
function printEnum(enumerator) {
  const list = enumerator.enumerators;
  const children = list.filter(Boolean).map(printEnumerator).join(`${EOL}  `).replace(/(\r?\n\s*){3,}/g, `${EOL + EOL}  `);
  return `
enum ${enumerator.name} {
  ${children}
}`;
}
function printEnumerator(enumerator) {
  switch (enumerator.type) {
    case 'enumerator':
      {
        const attrs = enumerator.attributes ? enumerator.attributes.map(printAttribute) : [];
        return [enumerator.name, ...attrs, enumerator.comment].filter(Boolean).join(' ');
      }
    case 'attribute':
      return printAttribute(enumerator);
    case 'comment':
      return printComment(enumerator);
    case 'break':
      return printBreak();
    default:
      throw new Error(`Unexpected enumerator type`);
  }
}
function printGenerator(generator) {
  const children = computeAssignmentFormatting(generator.assignments);
  return `
generator ${generator.name} {
  ${children}
}`;
}
function printObject(object) {
  const props = [...object.properties];
  let blockAttributeMoved = false;
  props.sort((a, b) => {
    if (a.type === 'attribute' && a.kind === 'object' && (b.type !== 'attribute' || b.type === 'attribute' && b.kind !== 'object')) {
      blockAttributeMoved = true;
      return 1;
    }
    if (b.type === 'attribute' && b.kind === 'object' && (a.type !== 'attribute' || a.type === 'attribute' && a.kind !== 'object')) {
      blockAttributeMoved = true;
      return -1;
    }
    return 0;
  });
  const attrIndex = props.findIndex(item => item.type === 'attribute' && item.kind === 'object');
  const needsSpace = !['break', 'comment'].includes(props[attrIndex - 1]?.type);
  if (blockAttributeMoved && needsSpace) {
    props.splice(attrIndex, 0, {
      type: 'break'
    });
  }
  const children = computePropertyFormatting(props);
  return `
${object.type} ${object.name} {
  ${children}
}`;
}
function printAssignment(node, keyLength = 0) {
  switch (node.type) {
    case 'comment':
      return printComment(node);
    case 'break':
      return printBreak();
    case 'assignment':
      return `${node.key.padEnd(keyLength)} = ${printValue(node.value)}`;
    default:
      throw new Error(`Unexpected assignment type`);
  }
}
function printProperty(node, nameLength = 0, typeLength = 0) {
  switch (node.type) {
    case 'attribute':
      return printAttribute(node);
    case 'field':
      return printField(node, nameLength, typeLength);
    case 'comment':
      return printComment(node);
    case 'break':
      return printBreak();
    default:
      throw new Error(`Unrecognized property type`);
  }
}
function printAttribute(attribute) {
  const args = attribute.args && attribute.args.length > 0 ? `(${attribute.args.map(printAttributeArg).filter(Boolean).join(', ')})` : '';
  const name = [attribute.name];
  if (attribute.group) name.unshift(attribute.group);
  return `${attribute.kind === 'field' ? '@' : '@@'}${name.join('.')}${args}`;
}
function printAttributeArg(arg) {
  return printValue(arg.value);
}
function printField(field, nameLength = 0, typeLength = 0) {
  const name = field.name.padEnd(nameLength);
  const fieldType = printFieldType(field).padEnd(typeLength);
  const attrs = field.attributes ? field.attributes.map(printAttribute) : [];
  const comment = field.comment;
  return [name, fieldType, ...attrs].filter(Boolean).join(' ').trim() + (comment ? ` ${comment}` : '');
}
function printFieldType(field) {
  const suffix = field.array ? '[]' : field.optional ? '?' : '';
  if (typeof field.fieldType === 'object') {
    switch (field.fieldType.type) {
      case 'function':
        {
          return `${printFunction(field.fieldType)}${suffix}`;
        }
      default:
        throw new Error(`Unexpected field type`);
    }
  }
  return `${field.fieldType}${suffix}`;
}
function printFunction(func) {
  const params = func.params ? func.params.map(printValue) : '';
  return `${func.name}(${params})`;
}
function printValue(value) {
  switch (typeof value) {
    case 'object':
      {
        if (value == null) return 'null';
        if ('type' in value) {
          switch (value.type) {
            case 'keyValue':
              return `${value.key}: ${printValue(value.value)}`;
            case 'function':
              return printFunction(value);
            case 'array':
              return `[${value.args != null ? value.args.map(printValue).join(', ') : ''}]`;
            case 'object':
              return value.properties.length > 0 ? `{ ${value.properties.map(printValue).join(', ')} }` : '{}';
            default:
              throw new Error(`Unexpected value type`);
          }
        }
        throw new Error(`Unexpected object value`);
      }
    default:
      return String(value);
  }
}
function computeAssignmentFormatting(list) {
  let pos = 0;
  const listBlocks = list.reduce((memo, current, index, arr) => {
    if (current.type === 'break') return memo;
    if (index > 0 && arr[index - 1].type === 'break') memo[++pos] = [];
    memo[pos].push(current);
    return memo;
  }, [[]]);
  const keyLengths = listBlocks.map(lists => lists.reduce((max, current) => Math.max(max, current.type === 'assignment' ? current.key.length : 0), 0));
  return list.map((item, index, arr) => {
    if (index > 0 && item.type !== 'break' && arr[index - 1].type === 'break') keyLengths.shift();
    return printAssignment(item, keyLengths[0]);
  }).filter(Boolean).join(`${EOL}  `).replace(/(\r?\n\s*){3,}/g, `${EOL + EOL}  `);
}
function computePropertyFormatting(list) {
  let pos = 0;
  const listBlocks = list.reduce((memo, current, index, arr) => {
    if (current.type === 'break') return memo;
    if (index > 0 && arr[index - 1].type === 'break') memo[++pos] = [];
    memo[pos].push(current);
    return memo;
  }, [[]]);
  const nameLengths = listBlocks.map(lists => lists.reduce((max, current) => Math.max(max, current.type === 'field' ? current.name.length : 0), 0));
  const typeLengths = listBlocks.map(lists => lists.reduce((max, current) => Math.max(max, current.type === 'field' ? printFieldType(current).length : 0), 0));
  return list.map((prop, index, arr) => {
    if (index > 0 && prop.type !== 'break' && arr[index - 1].type === 'break') {
      nameLengths.shift();
      typeLengths.shift();
    }
    return printProperty(prop, nameLengths[0], typeLengths[0]);
  }).filter(Boolean).join(`${EOL}  `).replace(/(\r?\n\s*){3,}/g, `${EOL + EOL}  `);
}

const findByType = (list, typeToMatch, options = {}) => {
  const [match, unexpected] = list.filter(findBy(typeToMatch, options));
  if (!match) return null;
  if (unexpected) throw new Error(`Found multiple blocks with [type=${typeToMatch}]`);
  return match;
};
const findAllByType = (list, typeToMatch, options = {}) => {
  return list.filter(findBy(typeToMatch, options));
};
const findBy = (typeToMatch, {
  name
} = {}) => block => {
  if (name != null) {
    const nameAttribute = typeToMatch === 'assignment' ? 'key' : 'name';
    if (!(nameAttribute in block)) return false;
    const nameMatches = typeof name === 'string' ? block[nameAttribute] === name : name.test(block[nameAttribute]);
    if (!nameMatches) return false;
  }
  return block.type === typeToMatch;
};

class ConcretePrismaSchemaBuilder {
  constructor(source = '') {
    this.schema = void 0;
    this._subject = void 0;
    this._parent = void 0;
    this.schema = getSchema(source);
  }
  print(options = {}) {
    return printSchema(this.schema, options);
  }
  getSchema() {
    return this.schema;
  }
  generator(name, provider = 'prisma-client-js') {
    const generator = this.schema.list.reduce((memo, block) => block.type === 'generator' && block.name === name ? block : memo, {
      type: 'generator',
      name,
      assignments: [{
        type: 'assignment',
        key: 'provider',
        value: `"${provider}"`
      }]
    });
    if (!this.schema.list.includes(generator)) this.schema.list.push(generator);
    this._subject = generator;
    return this;
  }
  drop(name) {
    const index = this.schema.list.findIndex(block => 'name' in block && block.name === name);
    if (index !== -1) this.schema.list.splice(index, 1);
    return this;
  }
  datasource(provider, url) {
    const datasource = {
      type: 'datasource',
      name: 'db',
      assignments: [{
        type: 'assignment',
        key: 'url',
        value: typeof url === 'string' ? `"${url}"` : {
          type: 'function',
          name: 'env',
          params: [`"${url.env}"`]
        }
      }, {
        type: 'assignment',
        key: 'provider',
        value: `"${provider}"`
      }]
    };
    const existingIndex = this.schema.list.findIndex(block => block.type === 'datasource');
    this.schema.list.splice(existingIndex, existingIndex !== -1 ? 1 : 0, datasource);
    this._subject = datasource;
    return this;
  }
  model(name) {
    const model = this.schema.list.reduce((memo, block) => block.type === 'model' && block.name === name ? block : memo, {
      type: 'model',
      name,
      properties: []
    });
    if (!this.schema.list.includes(model)) this.schema.list.push(model);
    this._subject = model;
    return this;
  }
  view(name) {
    const view = this.schema.list.reduce((memo, block) => block.type === 'view' && block.name === name ? block : memo, {
      type: 'view',
      name,
      properties: []
    });
    if (!this.schema.list.includes(view)) this.schema.list.push(view);
    this._subject = view;
    return this;
  }
  type(name) {
    const type = this.schema.list.reduce((memo, block) => block.type === 'type' && block.name === name ? block : memo, {
      type: 'type',
      name,
      properties: []
    });
    if (!this.schema.list.includes(type)) this.schema.list.push(type);
    this._subject = type;
    return this;
  }
  enum(name, enumeratorNames = []) {
    const e = this.schema.list.reduce((memo, block) => block.type === 'enum' && block.name === name ? block : memo, {
      type: 'enum',
      name,
      enumerators: enumeratorNames.map(name => ({
        type: 'enumerator',
        name
      }))
    });
    if (!this.schema.list.includes(e)) this.schema.list.push(e);
    this._subject = e;
    return this;
  }
  enumerator(value) {
    const subject = this.getSubject();
    if (!subject || !('type' in subject) || subject.type !== 'enum') {
      throw new Error('Subject must be a prisma enum!');
    }
    const enumerator = {
      type: 'enumerator',
      name: value
    };
    subject.enumerators.push(enumerator);
    this._parent = this._subject;
    this._subject = enumerator;
    return this;
  }
  getSubject() {
    return this._subject;
  }
  getParent() {
    return this._parent;
  }
  blockAttribute(name, args, params = {}) {
    let subject = this.getSubject();
    if (subject.type !== 'enum' && !isSchemaObject(subject)) {
      const parent = this.getParent();
      if (!isOneOfSchemaObjects(parent, ['model', 'view', 'type', 'enum'])) throw new Error('Subject must be a prisma model, view, or type!');
      subject = this._subject = parent;
    }
    const attributeArgs = (() => {
      const attributeArgs = [];
      const keyedArgs = args && !Array.isArray(args) && typeof args === 'object' ? {
        ...args,
        ...params
      } : params;
      if (typeof args === 'string') {
        attributeArgs.push({
          type: 'attributeArgument',
          value: `"${args}"`
        });
      } else if (Array.isArray(args)) {
        attributeArgs.push({
          type: 'attributeArgument',
          value: {
            type: 'array',
            args: args.map(arg => this.toValue(arg))
          }
        });
      }
      attributeArgs.push(...Object.entries(keyedArgs).map(([key, value]) => ({
        type: 'attributeArgument',
        value: {
          type: 'keyValue',
          key,
          value: this.toValue(value)
        }
      })));
      return attributeArgs;
    })();
    const property = {
      type: 'attribute',
      kind: 'object',
      name,
      args: attributeArgs
    };
    if (subject.type === 'enum') {
      subject.enumerators.push(property);
    } else {
      subject.properties.push(property);
    }
    return this;
  }
  attribute(name, args) {
    const parent = this.getParent();
    const subject = this.getSubject();
    if (!isOneOfSchemaObjects(parent, ['model', 'view', 'type', 'enum'])) {
      throw new Error('Parent must be a prisma model or view!');
    }
    if (!isSchemaField(subject)) {
      throw new Error('Subject must be a prisma field or enumerator!');
    }
    if (!subject.attributes) subject.attributes = [];
    const attribute = subject.attributes.reduce((memo, attr) => attr.type === 'attribute' && `${attr.group ? `${attr.group}.` : ''}${attr.name}` === name ? attr : memo, {
      type: 'attribute',
      kind: 'field',
      name
    });
    if (Array.isArray(args)) {
      if (args.length > 0) attribute.args = args.map(arg => ({
        type: 'attributeArgument',
        value: this.mapArg(arg)
      }));
    } else if (typeof args === 'object') {
      attribute.args = Object.entries(args).map(([key, value]) => ({
        type: 'attributeArgument',
        value: {
          type: 'keyValue',
          key,
          value: this.toValue(value)
        }
      }));
    }
    if (!subject.attributes.includes(attribute)) subject.attributes.push(attribute);
    return this;
  }
  mapArg(arg) {
    return typeof arg === 'string' ? arg : {
      type: 'function',
      name: arg.name,
      params: arg.function?.map(item => this.mapArg(item)) ?? []
    };
  }
  toValue(value) {
    if (Array.isArray(value)) {
      return {
        type: 'array',
        args: value.map(item => this.toValue(item))
      };
    }
    if (value === null || typeof value !== 'object') {
      return value;
    }
    if ('type' in value && typeof value.type === 'string' && ['array', 'function', 'object'].includes(value.type)) {
      return value;
    }
    return {
      type: 'object',
      properties: Object.entries(value).map(([key, item]) => ({
        type: 'keyValue',
        key,
        value: this.toValue(item)
      }))
    };
  }
  removeAttribute(name) {
    const parent = this.getParent();
    const subject = this.getSubject();
    if (!isSchemaObject(parent)) {
      throw new Error('Parent must be a prisma model or view!');
    }
    if (!isSchemaField(subject)) {
      throw new Error('Subject must be a prisma field!');
    }
    if (!subject.attributes) subject.attributes = [];
    subject.attributes = subject.attributes.filter(attr => !(attr.type === 'attribute' && attr.name === name));
    return this;
  }
  assignment(key, value) {
    const subject = this.getSubject();
    if (!subject || !('type' in subject) || !['generator', 'datasource'].includes(subject.type)) throw new Error('Subject must be a prisma generator or datasource!');
    function tap(subject, callback) {
      callback(subject);
      return subject;
    }
    const assignment = subject.assignments.reduce((memo, assignment) => assignment.type === 'assignment' && assignment.key === key ? tap(assignment, a => {
      a.value = `"${value}"`;
    }) : memo, {
      type: 'assignment',
      key,
      value: `"${value}"`
    });
    if (!subject.assignments.includes(assignment)) subject.assignments.push(assignment);
    return this;
  }
  findByType(typeToMatch, {
    within = this.schema.list,
    ...options
  }) {
    return findByType(within, typeToMatch, options);
  }
  findAllByType(typeToMatch, {
    within = this.schema.list,
    ...options
  }) {
    return findAllByType(within, typeToMatch, options);
  }
  blockInsert(statement) {
    let subject = this.getSubject();
    const allowed = ['datasource', 'enum', 'generator', 'model', 'view', 'type'];
    if (!subject || !('type' in subject) || !allowed.includes(subject.type)) {
      const parent = this.getParent();
      if (!parent || !('type' in parent) || !allowed.includes(parent.type)) {
        throw new Error('Subject must be a prisma block!');
      }
      subject = this._subject = parent;
    }
    switch (subject.type) {
      case 'datasource':
        {
          subject.assignments.push(statement);
          break;
        }
      case 'enum':
        {
          subject.enumerators.push(statement);
          break;
        }
      case 'generator':
        {
          subject.assignments.push(statement);
          break;
        }
      case 'model':
        {
          subject.properties.push(statement);
          break;
        }
    }
    return this;
  }
  break() {
    const lineBreak = {
      type: 'break'
    };
    return this.blockInsert(lineBreak);
  }
  comment(text, node = false) {
    const comment = {
      type: 'comment',
      text: `//${node ? '/' : ''} ${text}`
    };
    return this.blockInsert(comment);
  }
  schemaComment(text, node = false) {
    const comment = {
      type: 'comment',
      text: `//${node ? '/' : ''} ${text}`
    };
    this.schema.list.push(comment);
    return this;
  }
  field(name, fieldType = 'String') {
    let subject = this.getSubject();
    if (!isSchemaObject(subject)) {
      const parent = this.getParent();
      if (!isSchemaObject(parent)) throw new Error('Subject must be a prisma model or view or composite type!');
      subject = this._subject = parent;
    }
    const field = subject.properties.reduce((memo, block) => block.type === 'field' && block.name === name ? block : memo, {
      type: 'field',
      name,
      fieldType
    });
    if (!subject.properties.includes(field)) subject.properties.push(field);
    this._parent = subject;
    this._subject = field;
    return this;
  }
  removeField(name) {
    let subject = this.getSubject();
    if (!isSchemaObject(subject)) {
      const parent = this.getParent();
      if (!isSchemaObject(parent)) throw new Error('Subject must be a prisma model or view or composite type!');
      subject = this._subject = parent;
    }
    subject.properties = subject.properties.filter(field => !(field.type === 'field' && field.name === name));
    return this;
  }
  then(callback) {
    callback(this._subject);
    return this;
  }
}
function createPrismaSchemaBuilder(source) {
  return new ConcretePrismaSchemaBuilder(source);
}

function produceSchema(source, producer, options = {}) {
  const builder = createPrismaSchemaBuilder(source);
  producer(builder);
  return builder.print(options);
}

export { ConcretePrismaSchemaBuilder, PrismaParser, VisitorClassFactory, createPrismaSchemaBuilder, getSchema, printSchema, produceSchema };
